tensorzero-optimizers/src/gepa/pareto.rs

```diff
@@ -0,0 +1,2511 @@
+    /// 1. Checks for name collisions between new and existing variants (returns error if found)
+    /// 2. Merges new variants with existing frontier variants
+    /// 3. Runs GEPA's SELECTCANDIDATE algorithm on the combined set:
+    ///    a. For each datapoint, find instance-wise Pareto-optimal variants
+    ///    b. Build candidate set C as union of all instance-wise Pareto sets
+    ///    c. Filter candidates globally to remove dominated variants
+    /// 4. Updates frontier with filtered results and computes frequency weights
+    ///
+    /// This method validates that the provided validation scores match the frontier's
+    /// layout (datapoint_ids and evaluators), updates the frontier in-place, and prunes
+    /// validation scores to only contain current variants.
+    ///
+    /// # Arguments
+    /// * `new_variants` - Newly generated variants (mutations) to add to frontier
+    /// * `new_variant_scores` - Validation scores for the new variants
+    ///
+    /// # Returns
+    /// * `Result<(), Error>` - Ok if successful, Err if scores are invalid, collision, or layout mismatch
+    ///
+    /// # Errors
+    /// * Returns error if any new variant name already exists in the frontier
+    /// * Returns error if new_variant_scores is empty or all scores are None
+    /// * Returns error if new_variant_scores contains datapoints not in self.datapoint_ids
+    pub fn update(
+        &mut self,
+        new_variants: HashMap<String, UninitializedChatCompletionConfig>,
+        new_variant_scores: HashMap<String, VariantScores>,
+    ) -> Result<(), Error> {
+        tracing::info!(
+            "Updating Pareto frontier: merging {} new variants with {} existing variants",
+            new_variants.len(),
+            self.variants.len()
+        );
+
+        // Step 1: Check for name collisions
+        let collisions: Vec<&String> = new_variants
+            .keys()
+            .filter(|name| self.variants.contains_key(*name))
+            .collect();
+
+        if !collisions.is_empty() {
+            let collision_list = if collisions.len() <= 5 {
+                collisions
+                    .iter()
+                    .map(|s| s.as_str())
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            } else {
+                format!(
+                    "{} and {} more",
+                    collisions[..5]
+                        .iter()
+                        .map(|s| s.as_str())
+                        .collect::<Vec<_>>()
+                        .join(", "),
+                    collisions.len() - 5
+                )
+            };
+            return Err(Error::new(ErrorDetails::InternalError {
+                message: format!(
+                    "Variant name collision(s) detected: {collision_list}. New variants cannot have the same names as existing frontier variants."
+                ),
+            }));
+        }
```

Comment:
I don't mind having fancy error handling like this but please pull it into a separate method with an obvious name so this code is easy to read. Actually, the entire collision detection logic should be pulled out.